LUN-2505

bogdan-cornianu opened the pull request.

bogdan-cornianu committed:
Fix unexpected multiple ThumbnailOption in filer image plugin.

laurafeier commented on cmsplugin_filer_image/migrations/0002_delete_duplicate_thumbnailoptions.py:
```diff
@@ -0,0 +1,30 @@
+    ThumbnailOption = apps.get_model("cmsplugin_filer_image", "ThumbnailOption")
+    unique_fields = ["name", "width", "height", "crop", "upscale"]
+    duplicates = (ThumbnailOption.objects.values(*unique_fields)
+                  .order_by()
```
I think you can drop the order_by since it has no arguments.

laurafeier commented on cmsplugin_filer_image/migrations/0002_delete_duplicate_thumbnailoptions.py:
```diff
@@ -0,0 +1,30 @@
+    for duplicate in duplicates:
+        for thumbnail_option in ThumbnailOption.objects.filter(
+                **{x: duplicate[x] for x in unique_fields}
+        ).exclude(id=duplicate['max_id']):
```
I think you should save this queryset in a variable outside the for statement. It's much more readable(by dropping the weird round closing bracket in the for statement) and it's better for debugging purposes(think of the case where you might need to put a pdb to check the result of the queryset).

laurafeier commented on cmsplugin_filer_image/migrations/0002_delete_duplicate_thumbnailoptions.py:
```diff
@@ -0,0 +1,30 @@
+    """
+    ThumbnailOption = apps.get_model("cmsplugin_filer_image", "ThumbnailOption")
+    unique_fields = ["name", "width", "height", "crop", "upscale"]
+    duplicates = (ThumbnailOption.objects.values(*unique_fields)
```
Since you are annotating the max_id you should add it to the list of values otherwise I think it will execute other queries for each duplicate['max_id'].

bogdan-cornianu replied:
The query that gets executed for each duplicate['max_id'] is the correct one, there's no need to add it to the list of values.

laurafeier commented on cmsplugin_filer_image/migrations/0002_delete_duplicate_thumbnailoptions.py:
```diff
@@ -0,0 +1,30 @@
+                **{x: duplicate[x] for x in unique_fields}
+        ).exclude(id=duplicate['max_id']):
+            for filer_image in thumbnail_option.filerimage_set.all():
+                filer_image.thumbnail_option_id = duplicate['max_id']
```
I think you forgot to call save() for filer_image instance

bogdan-cornianu committed:
Code review improvements:
 - drop order_by.
 - extract query set in a variable.
 - remove max_id.
 - call save() for filer_image instance.

laurafeier commented on cmsplugin_filer_image/migrations/0002_delete_duplicate_thumbnailoptions.py:
```diff
@@ -0,0 +1,33 @@
+    for duplicate in duplicates:
+        qs = ThumbnailOption.objects.filter(
+            **{x: duplicate[x] for x in unique_fields})
+        id_to_keep = qs[:1][0].id
```
Since you need all the elements from the qs you can drop the slicing [:1], [1:] so that the only one query gets executed and not 2.

bogdan-cornianu committed:
Remove slicing.
